src/main/java/com/puppycrawl/tools/checkstyle/SarifLogger.java

```diff
@@ -0,0 +1,348 @@
+            );
+        }
+        else {
+            results.add(resultLineOnly
+                .replace(SEVERITY_LEVEL_PLACEHOLDER, renderSeverityLevel(event.getSeverityLevel()))
+                .replace(URI_PLACEHOLDER, event.getFileName())
+                .replace(LINE_PLACEHOLDER, Integer.toString(event.getLine()))
+                .replace(MESSAGE_PLACEHOLDER, escape(event.getMessage()))
+                .replace(RULE_ID_PLACEHOLDER, event.getViolation().getKey())
+            );
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * Following idea suppressions are false positives
+     *
+     * @noinspection DynamicRegexReplaceableByCompiledPattern
+     */
+    @Override
+    public void addException(AuditEvent event, Throwable throwable) {
+        final StringWriter stringWriter = new StringWriter();
+        final PrintWriter printer = new PrintWriter(stringWriter);
+        throwable.printStackTrace(printer);
+        if (event.getFileName() == null) {
+            results.add(resultErrorOnly
+                .replace(SEVERITY_LEVEL_PLACEHOLDER, renderSeverityLevel(event.getSeverityLevel()))
+                .replace(MESSAGE_PLACEHOLDER, escape(stringWriter.toString()))
+            );
+        }
+        else {
+            results.add(resultFileOnly
+                .replace(SEVERITY_LEVEL_PLACEHOLDER, renderSeverityLevel(event.getSeverityLevel()))
+                .replace(URI_PLACEHOLDER, event.getFileName())
+                .replace(MESSAGE_PLACEHOLDER, escape(stringWriter.toString()))
+            );
+        }
+    }
+
+    @Override
+    public void fileStarted(AuditEvent event) {
+        // No need to implement this method in this class
+    }
+
+    @Override
+    public void fileFinished(AuditEvent event) {
+        // No need to implement this method in this class
+    }
+
+    /**
+     * Render the severity level into SARIF severity level.
+     *
+     * @param severityLevel the Severity level.
+     * @return the rendered severity level in string.
+     */
+    private static String renderSeverityLevel(SeverityLevel severityLevel) {
+        final String renderedSeverityLevel;
+        switch (severityLevel) {
+            case IGNORE:
+                renderedSeverityLevel = "none";
+                break;
+            case INFO:
+                renderedSeverityLevel = "note";
+                break;
+            case WARNING:
+                renderedSeverityLevel = "warning";
+                break;
+            case ERROR:
+            default:
+                renderedSeverityLevel = "error";
+                break;
+        }
+        return renderedSeverityLevel;
+    }
+
+    /**
+     * Escape \b, \f, \n, \r, \t, \", \\ and U+0000 through U+001F.
+     * Reference: https://www.ietf.org/rfc/rfc4627.txt - 2.5. Strings
+     *
+     * @param value the value to escape.
+     * @return the escaped value if necessary.
+     */
+    public static String escape(String value) {
```

Comment:
there is a method, I think it can be reused
https://github.com/checkstyle/checkstyle/blob/b4bdaa5fbe5d9b62eb6900b881a642f0af711e32/src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java#L249-L288

Reply:
XML encoding and JSON encoding are different, so keeping them separated reduces future maintenance cost.